### PR TITLE
Add Text2ImagePipeline to llm bench

### PR DESCRIPTION
**Pull request by sbalandi**
Ticket: CVS-156418

### CODE REVIEW

File: tools/llm_bench/benchmark.py [hunk 1]
```diff
@@ -129,6 +129,13 @@ def get_argprser():
     parser.add_argument('-od', '--output_dir', help='Save the input text and generated text, images to files')
     llm_bench_utils.model_utils.add_stateful_model_arguments(parser)
     parser.add_argument("--genai", action="store_true", help="Use OpenVINO GenAI optimized pipelines for benchmarking")
+    parser.add_argument(
+        "--lora",
+        nargs='*',
+        required=False,
+        default=None,
+        help="Path to LoRA adapters for using OpenVINO GenAI Text2Image optimized pipelines with LoRA for benchmarking")
```
Comment (ilya-lavrenov) [inline tools/llm_bench/benchmark.py:137]: LoRA can be applied for LLM pipelines as well (and for VLM and Whisper in next releases)
Should it be a generic parameter? not only for Text2Image

File: tools/llm_bench/llm_bench_utils/ov_utils.py [hunk 2]
```diff
@@ -255,12 +255,47 @@ def create_image_gen_model(model_path, device, **kwargs):
     if not Path(model_path).exists():
         raise RuntimeError(f'==Failure ==: model path:{model_path} does not exist')
     else:
+        if kwargs.get("genai", False) and is_genai_available(log_msg=True):
+            if model_class not in [OV_MODEL_CLASSES_MAPPING[default_model_type], OV_MODEL_CLASSES_MAPPING["mpt"], OV_MODEL_CLASSES_MAPPING["chatglm"]]:
+                log.warning("OpenVINO GenAI based benchmarking is not available for {model_type}. Will be switched to default bencmarking")
```
Comment (ilya-lavrenov) [inline tools/llm_bench/llm_bench_utils/ov_utils.py:260]:
```suggestion
                log.warning("OpenVINO GenAI based benchmarking is not available for {model_type}. Will be switched to default benchmarking")
```

File: src/python/py_image_generation_pipelines.cpp [hunk 3]
```diff
@@ -275,7 +275,7 @@ void init_image_generation_pipelines(py::module_& m) {
             (text2image_generate_docstring + std::string(" \n ")).c_str()
         );
 
-    auto image_generation_scheduler = py::class_<ov::genai::Scheduler>(m, "Scheduler", "Scheduler for image generation pipelines.")
+    auto image_generation_scheduler = py::class_<ov::genai::Scheduler, std::shared_ptr<ov::genai::Scheduler>>(m, "Scheduler", "Scheduler for image generation pipelines.")
         .def(py::init<>())
```
Comment (ilya-lavrenov) [inline src/python/py_image_generation_pipelines.cpp:279]: @popovaan 
can we drop this ctor?

File: tools/llm_bench/llm_bench_utils/ov_utils.py [hunk 4]
```diff
@@ -255,12 +255,47 @@ def create_image_gen_model(model_path, device, **kwargs):
...
+                adapter_config = openvino_genai.AdapterConfig()
+                adapter = openvino_genai.Adapter(kwargs['lora'][idx])
+                alpha = float(kwargs['lora_alphas'][idx])
+                adapter_config.add(adapter, alpha)
+            ov_config['adapters'] = adapter_config
+
+    start = time.perf_counter()
+    tokenizer = None
+    if (model_path / 'tokenizer').exists():
+        tokenizer = AutoTokenizer.from_pretrained(model_path / 'tokenizer')
```
Comment (eaidova) [inline tools/llm_bench/llm_bench_utils/ov_utils.py:294]: I do not think that it is needed, tokenization is part of pipeline and image generation performance usually does not depend from len of input prompt. Also I do not thank that is is valid and suitable for all pipelines as pipeline may have several tokenizers with different behaviour (in llm pipeline it was used for backward compatibility to be able check that tokenized text is the same like in original

Comment (eaidova) [reply tools/llm_bench/llm_bench_utils/ov_utils.py:294]: BTW, why tokenizer creatiion time included in pipeline initialization time?

File: tools/llm_bench/task/image_generation.py [hunk 5]
```diff
@@ -95,13 +95,85 @@ def run_image_generation(image_param, num, image_id, pipe, args, iter_data_list,
     stable_diffusion_hook.clear_statistics()
 
 
+def run_image_generation_genai(image_param, num, image_id, pipe, args, iter_data_list, proc_id, mem_consumption, tokenizer):
+    set_seed(args['seed'])
+    input_text = image_param['prompt']
+    image_width = image_param.get('width', DEFAULT_IMAGE_WIDTH)
+    image_height = image_param.get('height', DEFAULT_IMAGE_HEIGHT)
+    nsteps = image_param.get('steps', DEFAULT_INFERENCE_STEPS if 'lcm' not in args["model_name"] else LCM_DEFAULT_INFERENCE_STEPS)
```
Comment (eaidova) [inline tools/llm_bench/task/image_generation.py:103]: probably this part can be generalized in future (there are except lcm, there are also turbo models that may work with 1 step and hyper 2-8)

File: tools/llm_bench/task/image_generation.py [hunk 6]
```diff
@@ -95,13 +95,85 @@ def run_image_generation(image_param, num, image_id, pipe, args, iter_data_list,
...
+    max_rss_mem_consumption = ''
+    max_uss_mem_consumption = ''
+    max_shared_mem_consumption = ''
+    if (args['mem_consumption'] == 1 and num == 0) or args['mem_consumption'] == 2:
+        mem_consumption.start_collect_memory_consumption()
+    additional_args = {}
+    if guidance_scale is not None:
+        additional_args["guidance_scale"] = guidance_scale
+    else:
+        if 'lcm-sdxl' in args['model_type']:
```
Comment (eaidova) [inline tools/llm_bench/task/image_generation.py:119]: why only sdxl?

File: tools/llm_bench/task/image_generation.py [hunk 7]
```diff
@@ -95,13 +95,85 @@ def run_image_generation(image_param, num, image_id, pipe, args, iter_data_list,
...
+    )
+    iter_data_list.append(iter_data)
+    metrics_print.print_metrics(
+        num,
+        iter_data,
+        warm_up=(num == 0),
+        max_rss_mem=max_rss_mem_consumption,
+        max_shared_mem=max_shared_mem_consumption,
+        max_uss_mem=max_uss_mem_consumption,
+        stable_diffusion=None,
```
Comment (eaidova) [inline tools/llm_bench/task/image_generation.py:164]: why? Is it possibly to get detaild perf metrics from text2image pipeline?

File: tools/llm_bench/task/image_generation.py [hunk 8]
```diff
@@ -120,6 +192,11 @@ def run_image_generation_benchmark(model_path, framework, device, args, num_iter
         raise RuntimeError('==Failure prompts is empty ==')
     log.info(f'Benchmarking iter nums(exclude warm-up): {num_iters}, prompt nums: {len(image_list)}, prompt idx: {prompt_idx_list}')
 
+    if not use_genai:
+        text_gen_fn = run_image_generation
```
Comment (eaidova) [inline tools/llm_bench/task/image_generation.py:196]: very misleading naming....
```suggestion
        image_gen_fn = run_image_generation
```

File: tools/llm_bench/task/image_generation.py [hunk 9]
```diff
@@ -128,7 +205,7 @@ def run_image_generation_benchmark(model_path, framework, device, args, num_iter
             for image_id, image_param in enumerate(image_list):
                 p_idx = prompt_idx_list[image_id]
                 iter_timestamp[num][p_idx]['start'] = datetime.datetime.now().isoformat()
-                run_image_generation(image_param, num, prompt_idx_list[image_id], pipe, args, iter_data_list, proc_id, mem_consumption)
+                text_gen_fn(image_param, num, prompt_idx_list[image_id], pipe, args, iter_data_list, proc_id, mem_consumption, tokenizer)
```
Comment (eaidova) [inline tools/llm_bench/task/image_generation.py:208]:
```suggestion
                image_gen_fn(image_param, num, prompt_idx_list[image_id], pipe, args, iter_data_list, proc_id, mem_consumption, tokenizer)
```

File: tools/llm_bench/task/image_generation.py [hunk 10]
```diff
@@ -137,7 +214,7 @@ def run_image_generation_benchmark(model_path, framework, device, args, num_iter
             p_idx = prompt_idx_list[image_id]
             for num in range(num_iters + 1):
                 iter_timestamp[num][p_idx]['start'] = datetime.datetime.now().isoformat()
-                run_image_generation(image_param, num, p_idx, pipe, args, iter_data_list, proc_id, mem_consumption)
+                text_gen_fn(image_param, num, p_idx, pipe, args, iter_data_list, proc_id, mem_consumption, tokenizer)
```
Comment (eaidova) [inline tools/llm_bench/task/image_generation.py:217]:
```suggestion
                image_gen_fn(image_param, num, p_idx, pipe, args, iter_data_list, proc_id, mem_consumption, tokenizer)
```

File: tools/llm_bench/task/image_generation.py (hunk 6 above)
Comment (eaidova) [reply tools/llm_bench/task/image_generation.py:119]: I also belive that these parameters like guidance_scale, num steps e.g. are common for both inference pipelines, can we move that into utils and reuse for both?

File: tools/llm_bench/llm_bench_utils/ov_utils.py [hunk 11]
```diff
@@ -255,12 +255,47 @@ def create_image_gen_model(model_path, device, **kwargs):
     if not Path(model_path).exists():
         raise RuntimeError(f'==Failure ==: model path:{model_path} does not exist')
     else:
+        if kwargs.get("genai", False) and is_genai_available(log_msg=True):
+            if model_class not in [OV_MODEL_CLASSES_MAPPING[default_model_type], OV_MODEL_CLASSES_MAPPING["mpt"], OV_MODEL_CLASSES_MAPPING["chatglm"]]:
```
Comment (eaidova) [inline tools/llm_bench/llm_bench_utils/ov_utils.py:259]: mpt, chatglm - llms and can not be used in image generation, default model type is decoder also for text generation, seems that you do not need this condition

File: tools/llm_bench/task/image_generation.py (hunk 5 above)
Comment (ilya-lavrenov) [reply tools/llm_bench/task/image_generation.py:103]: why not to rely on default `inference_steps` defined in pipeline itself?

File: tools/llm_bench/task/image_generation.py (hunk 7 above)
Comment (ilya-lavrenov) [reply tools/llm_bench/task/image_generation.py:164]: no, text2image does not currently provide perf metrics. You can only measure `generate` itself

File: tools/llm_bench/task/image_generation.py (hunk 5 above)
Comment (eaidova) [reply tools/llm_bench/task/image_generation.py:103]: it is not always correct, e.g. adding lcm lora to pipeline keep pipeline default steps like in original while it should be manually reduced

File: src/python/py_image_generation_pipelines.cpp (hunk 3 above)
Comment (popovaan) [reply src/python/py_image_generation_pipelines.cpp:279]: Yes, made a PR to drop it: https://github.com/openvinotoolkit/openvino.genai/pull/1136

**Comment by eaidova** [comment]
@sbalandi could you please also add test run with --genai option in https://github.com/openvinotoolkit/openvino.genai/blob/master/.github/workflows/llm_bench-python.yml#L73

**Comment by sbalandi** [comment]
dreamlike-art/dreamlike-anime-1.0 is added @eaidova could you please confirm if it's okay, or too big ?
segmind/tiny-sd has unsupported type of scheduler `DPMSolverMultistepScheduler`

I have also tried segmind/tiny-sd-mxfinetune , this model has supported type of scheduler, but it fails in [scheduler.cpp](https://github.com/openvinotoolkit/openvino.genai/blob/master/src/cpp/src/image_generation/schedulers/scheduler.cpp#L22) on creating pipeline, because scheduler.json has value [`-Infinity`](https://huggingface.co/segmind/tiny-sd-mxfinetune/blob/main/scheduler/scheduler_config.json#L11) . @ilya-lavrenov is it expected ? should we handle it ?

**Comment by ilya-lavrenov** [comment]
> [dreamlike-art/dreamlike-anime-1.0](https://huggingface.co/dreamlike-art/dreamlike-anime-1.0) is added @eaidova could you please confirm if it's okay, or too big ? segmind/tiny-sd has unsupported type of scheduler `DPMSolverMultistepScheduler`
> 
> I have also tried segmind/tiny-sd-mxfinetune , this model has supported type of scheduler, but it fails in [scheduler.cpp](https://github.com/openvinotoolkit/openvino.genai/blob/master/src/cpp/src/image_generation/schedulers/scheduler.cpp#L22) on creating pipeline, because scheduler.json has value [`-Infinity`](https://huggingface.co/segmind/tiny-sd-mxfinetune/blob/main/scheduler/scheduler_config.json#L11) . @ilya-lavrenov is it expected ? should we handle it ?

According to https://jsonlint.com/, it's invalid XML and I don't want to create WAs on GenAI for such cases (or only for popular models)

**Comment by sbalandi** [comment]
> > [dreamlike-art/dreamlike-anime-1.0](https://huggingface.co/dreamlike-art/dreamlike-anime-1.0) is added @eaidova could you please confirm if it's okay, or too big ? segmind/tiny-sd has unsupported type of scheduler `DPMSolverMultistepScheduler`
> > I have also tried segmind/tiny-sd-mxfinetune , this model has supported type of scheduler, but it fails in [scheduler.cpp](https://github.com/openvinotoolkit/openvino.genai/blob/master/src/cpp/src/image_generation/schedulers/scheduler.cpp#L22) on creating pipeline, because scheduler.json has value [`-Infinity`](https://huggingface.co/segmind/tiny-sd-mxfinetune/blob/main/scheduler/scheduler_config.json#L11) . @ilya-lavrenov is it expected ? should we handle it ?
> 
> According to https://jsonlint.com/, it's invalid XML and I don't want to create WAs on GenAI for such cases (or only for popular models)
> 
> > unsupported type of scheduler DPMSolverMultistepScheduler
> 
> You can override scheduler to support one via `pipe.set_scheduler` method.

okay, thank you

File: .github/workflows/llm_bench-python.yml [hunk 12]
```diff
@@ -74,6 +74,14 @@ jobs:
       run: |
         optimum-cli export openvino --model segmind/tiny-sd --trust-remote-code --weight-format fp16 ./ov_models/tiny-sd/pytorch/dldt/FP16/
         python ./tools/llm_bench/benchmark.py -m ./ov_models/tiny-sd/pytorch/dldt/FP16/ -pf ./tools/llm_bench/prompts/stable-diffusion.jsonl -d cpu -n 1
+    - name: Test tiny-stable-diffusion on Linux with GenAI
```
Comment (eaidova) [inline .github/workflows/llm_bench-python.yml:77]:
```suggestion
    - name: Test dreamlike-anime on Linux with GenAI
```

File: .github/workflows/llm_bench-python.yml [hunk 13]
```diff
@@ -74,6 +74,14 @@ jobs:
...
+      run: |
+        optimum-cli export openvino --model dreamlike-art/dreamlike-anime-1.0 --task stable-diffusion --weight-format fp16 ov_models/dreamlike-art-dreamlike-anime-1.0/FP16
+        python ./tools/llm_bench/benchmark.py -m ./ov_models/dreamlike-art-dreamlike-anime-1.0/FP16/ -pf ./tools/llm_bench/prompts/stable-diffusion.jsonl -d cpu -n 1 --genai
+    - name: Test tiny-stable-diffusion on Linux with GenAI and LoRA
```
Comment (eaidova) [inline .github/workflows/llm_bench-python.yml:81]:
```suggestion
    - name: Test dreamlike-anime on Linux with GenAI and LoRA
```

**Comment by eaidova** [comment]
> [dreamlike-art/dreamlike-anime-1.0](https://huggingface.co/dreamlike-art/dreamlike-anime-1.0) is added @eaidova could you please confirm if it's okay, or too big ? segmind/tiny-sd has unsupported type of scheduler `DPMSolverMultistepScheduler`
> 
> I have also tried segmind/tiny-sd-mxfinetune , this model has supported type of scheduler, but it fails in [scheduler.cpp](https://github.com/openvinotoolkit/openvino.genai/blob/master/src/cpp/src/image_generation/schedulers/scheduler.cpp#L22) on creating pipeline, because scheduler.json has value [`-Infinity`](https://huggingface.co/segmind/tiny-sd-mxfinetune/blob/main/scheduler/scheduler_config.json#L11) . @ilya-lavrenov is it expected ? should we handle it ?

@sbalandi for me, any variant of model is ok, until it works in GHA

**Comment by ilya-lavrenov** [comment]
port to 2024.5 ?

**Comment by eaidova** [comment]
> port to 2024.5 ?
these changes present in release
image

**Pull request merged**